Towards an API to apply forces to an MBT model.

amcastro-tri opened the pull request:
This PR introduces a little class, `MultibodyForcing`, that simply stores the collection of generalized forces and body forces to a multibody system. By passing around objects to this type to MBT methods, we will be able to abstract the user from having to think about entries into arrays of forces. This will help introduce (in future PR) an API to apply joint as well as body forces in a homogeneous way.

This first PR uses this little class to cleanup force elements API (which by the way still requires further clean up after caching lands...)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7731)
<!-- Reviewable:end -->

amcastro-tri commented:
+@edrumwri for feature review please.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2XCY0MeUAEiRQRj74Y:b-a98c4g)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
-@edrumwri 

+@seancurtis-tri for feature review please.

---

Review status: 0 of 12 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2a-lgwSfFiA-XyuUPs:bpannl4)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Three things:

1. Some small cosmetic stuff,
2. Some API questions, and
3. FTR,  "Forcing" feels like an odd term. It seems that this should be called `MultibodyForce`. Here's my reasoning: a `SpatialForce` is, essentially, a union of two kinds of forces. The current `MultibodyForcing` is likewise a union of two kinds of sets of forces (spatial and generalized). In the first case, we called the union of two forces another force, then it seems we can define this union as another kind of force (whose dimensionality depends on the number of bodies and dofs in the multibody tree). If we do this, it seems to make more sense to talk about adding forces, etc.  Thoughts?

---

Reviewed 2 of 19 files at r1, 11 of 17 files at r2.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[examples/multibody/pendulum/pendulum_plant.cc, line 287 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aYkqQEg1_AuLqLXsk:-L2aYkqRGKpXEV9KunAb:b-blowv9) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/examples/multibody/pendulum/pendulum_plant.cc#L287)):*
> ```cpp
>   model_->CalcInverseDynamics(
>       context, pc, vc, vdot,
>       F_BBo_W_array, tau_array,
> ```

BTW Will this eventually just take the `MultibodyForcing` object as an input?

---

*[multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 309 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aZ8Yvd41ItptK6X8G:-L2aZ8Yvd41ItptK6X8H:bw1ud6o) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L309)):*
BTW Awkward language, recommend:

> Adds the previously included effect of gravity into `forcing`.

---

*[multibody/multibody_tree/multibody_forcing.h, line 16 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aL-sj0-oP87n9ALOX:-L2aL-sj0-oP87n9ALOY:ba3qhmk) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L16)):*
> ```objc
> 
> /// A class to hold the external forcing applied to a MultibodyTree system.
> /// External forcing might include generalized forces at joints as well as
> ```

BTW s/might/can

---

*[multibody/multibody_tree/multibody_forcing.h, line 34 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aLIgg42yIecYveYTU:-L2aLIgg42yIecYveYTV:b-6irlww) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L34)):*
> ```objc
> 
>   /// Constructs a forcing object compatible with `model`. Forcing is
>   /// initialized to zero, meaning no forces are applied to `model`.
> ```

BTW No documentation on the restrictions of the model (e.g., must be finalized.)

---

*[multibody/multibody_tree/multibody_forcing.h, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aT1ErcYLZhMi-j6Z5:-L2aT1ErcYLZhMi-j6Z6:b-cr1jj) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L48)):*
> ```objc
>   /// Returns the number of mobilities for which this forcing applies.
>   /// Determined at construction from the given model MultibodyTree object.
>   int num_mobilities() const {
> ```

Should this be mobilizers or joints? It's not clear at what level this API lives. The class documentation refers to forces "at joints", but then this is mobilizers. It seems that the language should be unified so that it's clear whether this abstraction belongs at the mobilizer level or the joint level.

---

*[multibody/multibody_tree/multibody_forcing.h, line 64 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aTfdsFg4_x_Pj9h60:-L2aTfdsFg4_x_Pj9h61:b-jmgqvg) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L64)):*
> ```objc
>   /// Returns a constant reference to the vector of applied spatial forces
>   /// F_B_W on each body B in the model, expressed in the world frame W.
>   /// @note Entries are ordered by BodyNodeIndex.
> ```

BTW For my own edification, is it well understood *where* the force is applied? Does "on the body" mean "at the center of mass of the body" or some such thing?

---

*[multibody/multibody_tree/multibody_forcing.h, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aU1oMkd0ublsNOEjP:-L2aU1oMkd0ublsNOEjQ:bc6bolv) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L80)):*
> ```objc
>   /// MultibodyTree model.
>   /// @returns true if `this` forcing is compatible with `model`.
>   bool CheckInvariants(const MultibodyTree<T>& model) const;
> ```

BTW Is this public for a reason other than testing?

BTW Also, is it worth noting that this is a liberal definition of "compatible"? I.e., it tests that its appropriately sized, but there's no guarantees that the semantics of individual forces are "right".

---

*[multibody/multibody_tree/multibody_tree.h, line 903 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aWG8cyNDqWdNtxXlC:-L2aWG8cyNDqWdNtxXlD:b-hgxc8i) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_tree.h#L903)):*
> ```objc
>   ///   A velocity kinematics cache object already updated to be in sync with
>   ///   `context`.
>   /// @param[out] forcing
> ```

BTW With the change in API, you no longer need this as an *out* parameter. You could simply return a `MultibodyForcing` instance. 

---

*[multibody/multibody_tree/test/double_pendulum_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aWlzV8IE6fyLNZmmf:-L2aWlzV8IE6fyLNZmmg:bfzpyg5) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/double_pendulum_test.cc#L1)):*
> ```cpp
> // clang-format: off
> ```

FYI I think this will turn into a merge headache; lots of renamed to a then renamed to b. Keep an eye out for this one.

---

*[multibody/multibody_tree/test/multibody_forcing_test.cc, line 10 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aZg9u2n-222nobBnP:-L2aZg9u2n-222nobBnQ:b-jvenzm) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/multibody_forcing_test.cc#L10)):*
> ```cpp
> #include "drake/common/eigen_types.h"
> #include "drake/common/test_utilities/eigen_matrix_compare.h"
> #include "drake/multibody/multibody_tree/multibody_tree.h"
> ```

BTW I'm betting you have include files out of order. Not sure how it passed CI.

---

*[multibody/multibody_tree/test/multibody_forcing_test.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aZv3RJDUBK0sbHvGz:-L2aZv3RJDUBK0sbHvH-:bbz8bkd) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/multibody_forcing_test.cc#L51)):*
> ```cpp
>   // Forcing object should be compatible with the original model.
>   EXPECT_TRUE(forcing.CheckInvariants(model_));
>   EXPECT_EQ(forcing.num_bodies(), model_.get_num_bodies());
> ```

BTW These two tests feel redundant -- they are already tested by `CheckInvariants()`, yes?

---

*[multibody/multibody_tree/test/multibody_forcing_test.cc, line 57 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2a_3O6YtgH3u8b141R:-L2a_3O6YtgH3u8b141S:b-3j92te) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/multibody_forcing_test.cc#L57)):*
> ```cpp
> 
>   for (const SpatialForce<double>& F : forcing.body_forces()) {
>     EXPECT_TRUE(F.IsApprox(SpatialForce<double>::Zero(), kEpsilon));
> ```

BTW Surely this can (and should) be compared with epsilon = zero.

---

*[multibody/multibody_tree/test/multibody_forcing_test.cc, line 105 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aauD-hgawTCCaKqHl:-L2aauD-hgawTCCaKqHm:bb0ji2v) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/multibody_forcing_test.cc#L105)):*
> ```cpp
> 
>   // Set to zero and assess the result:
>   forcing1.SetZero();
> ```

BTW Worth adding a `MultibodyForcing::IsZero() const` method?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2aU9bmN2_IMXfPaDnt:bgkaylm)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 7 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[examples/multibody/pendulum/pendulum_plant.cc, line 287 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aYkqQEg1_AuLqLXsk:-L2amP-TYo1LPEgkD7zE:b3l0jx1) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/examples/multibody/pendulum/pendulum_plant.cc#L287)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Will this eventually just take the `MultibodyForcing` object as an input?
</blockquote></details>

BTW, yes. I didn't introduce the change just yet to keep the PR small. Also this will get wrapped on a method that doesn't take cache entries once caching lands so I might just wait until that happens to introduce the final API.

---

*[multibody/multibody_tree/multibody_forcing.h, line 48 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aT1ErcYLZhMi-j6Z5:-L2aoZW4WNrmac3RXp4C:b-uqro95) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L48)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Should this be mobilizers or joints? It's not clear at what level this API lives. The class documentation refers to forces "at joints", but then this is mobilizers. It seems that the language should be unified so that it's clear whether this abstraction belongs at the mobilizer level or the joint level.
</blockquote></details>

changed to `num_velocities` for consistency with other places.

---

*[multibody/multibody_tree/multibody_forcing.h, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aU1oMkd0ublsNOEjP:-L2ap3jQBC1tjcviMPbf:b-9liqei) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L80)):*
> Is this public for a reason other than testing?

It is used within MBT methods to ensure forcing objects are compatible with the model. See for instance `CalcAndAddForceContribution()`.

> Also, is it worth noting that this is a liberal definition of...

We could rename it `CheckSizes()` if you think that'd be more appropriate. If the tree eventually had an id (I think we talked about that before) then we could check this forcing is for that particular MBT. This API would allow that without changes (or any other extra checks we might decide to add in the future).

---

*[multibody/multibody_tree/test/double_pendulum_test.cc, line 1 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aWlzV8IE6fyLNZmmf:-L2aqibuvebCYHp4LYiA:bx9l08p) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/double_pendulum_test.cc#L1)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI I think this will turn into a merge headache; lots of renamed to a then renamed to b. Keep an eye out for this one.
</blockquote></details>

ah.. I think that's because I didn't merge the previous PR you reviewed yet. I'll double check after merging that one. Thanks

---

*[multibody/multibody_tree/test/multibody_forcing_test.cc, line 10 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aZg9u2n-222nobBnP:-L2as5p_Smn7ypMLBPmv:bpwrehj) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/multibody_forcing_test.cc#L10)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I'm betting you have include files out of order. Not sure how it passed CI.
</blockquote></details>

BTW I know, I somehow eliminated the `BUILD.bazel` change where I add this test :-/. Thanks!

---

*[multibody/multibody_tree/test/multibody_forcing_test.cc, line 51 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aZv3RJDUBK0sbHvGz:-L2ar72jJtqg0WsC_OdZ:b-eji8l1) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/test/multibody_forcing_test.cc#L51)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW These two tests feel redundant -- they are already tested by `CheckInvariants()`, yes?
</blockquote></details>

BTW Yes, but they are there to test the API, not the actual already tested sizes as you mention. Added comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
I addressed 1 and 2 @seancurtis-tri.
Regarding 3, I actually purposely preferred not to use `MultibodyForce` to avoid any confusion with `SpatialForce`. While `SpatialForce` is an actual math object, `MultibodyForcing` is just a convenience container that will help hide from users how exactly forces are applied (users only care about applying joint torques or body forces in some frame but now how they are applied). Therefore the naming I chose. Not super strong about it and I could be convinced otherwise if there are other views about this.

---

Review status: 7 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2asvEBQA884xkiMZ8R:b-5p5gyc)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I realize, in retrospect, that `MultibodyForce` is bad. I did a quick straw poll with @rcory and @hongkai-dai. Everyone agrees that `MultibodyForcing` doesn't communicate much. They were unable to come up with an alternative and in the effort made the observation that lumping the two quantities together (spatial and generalized forces) is strange and confusing. They never collect them together, so they have no name for them. So, my question is, this seems to be a PR about simplifying parameter lists -- maybe we should confirm that we're not exchanging *clarity* for more compact interfaces. I'd suggest pinging some of the roboticists at your end. I'll also ping @sherm1, @edrumwri, and @mitiguy  when they get back.

---

Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/multibody_tree/multibody_forcing.h, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aU1oMkd0ublsNOEjP:-L2b2KI0i6MpSxNm9FSy:b-qxxvpg) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forcing.h#L80)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

> Is this public for a reason other than testing?

It is used within MBT methods to ensure forcing objects are compatible with the model. See for instance `CalcAndAddForceContribution()`.

> Also, is it worth noting that this is a liberal definition of...

We could rename it `CheckSizes()` if you think that'd be more appropriate. If the tree eventually had an id (I think we talked about that before) then we could check this forcing is for that particular MBT. This API would allow that without changes (or any other extra checks we might decide to add in the future).
</blockquote></details>

I'd be content if it was simply documented. In other words, you have a definition of compatible that isn't obvious. Go ahead and define it and I'm good.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2azLEBy0luCM9DCHIH:ba8lzg6)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
That is because in RBT land you know all the internal details (of course super good if you are doing something specialized). Most peopled don't do anything specialized, they just want to apply say a torque on a Joint regardless of what the underlying model for that joint is (a mobilizer, a constraint, a stiff set of springs). Therefore it is not clear for a user, unless you dig into the internals, how that forcing should be applied; through either generalized forces or body forces.
Ultimately, the idea would be a user can do something like (say you have a revolute joint):
```c++
MultibodyTree<double> model;
... create your model, do stuff...
MultibodyForcing<double> forcing(model);
double torque = 2.0;
joint->AddInTorque(context, torque, &forcing);
... do more stuff...
CalcForwardDynamics(context, forcing);
... do more stuff...
SpatialAcceleration<double> A_WB = forearm->EvalSpatialAcceleration(context);
```
that is, this offers a higher level API to apply your forces for those of who are not interested in knowing how exactly that happens (I'd assume that'd be a lot of people). 
Let me know what you guys think about this. 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2b3l1drt0ENn1Y-EWk:bggxq1t)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I pinged the guys. We converged on happiness with the class (and the API), but felt that the name `MultibodyForces` (note plural), as the preferred name.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2bIuqI7_wdB_k-M1bN:b-mmu62d)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
As requested, I made the rename to `MultibodyForces`.

---

Review status: 0 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/multibody_forces.h, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aU1oMkd0ublsNOEjP:-L2f-9UJWeaoS7VUF9CW:bljxiv2) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forces.h#L80)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I'd be content if it was simply documented. In other words, you have a definition of compatible that isn't obvious. Go ahead and define it and I'm good.
</blockquote></details>

I rephrased this a bit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2evQh6tdNmr_ixwfBD:bwgjeri)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with two cosmetic comments.

---

Reviewed 10 of 16 files at r4.
Review status: 9 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[examples/multibody/pendulum/pendulum_plant.cc, line 256 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2f231GP_mlk5o_vBDk:-L2f231GP_mlk5o_vBDl:bwks4xt) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/examples/multibody/pendulum/pendulum_plant.cc#L256)):*
> ```cpp
>   MatrixX<T> M(nv, nv);
>   // Forces:
>   MultibodyForces<T> forcing(*model_);
> ```

BTW (meta) with the class name change, did you want a variable name change as well?

---

*[multibody/multibody_tree/multibody_forces.h, line 80 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2aU1oMkd0ublsNOEjP:-L2f2bzTNmzEGdhTjJt1:bbshyxe) ([raw file](https://github.com/robotlocomotion/drake/blob/a14b3540bc4c6d067d62635bf2d745ecc654dedc/multibody/multibody_tree/multibody_forces.h#L80)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I rephrased this a bit.
</blockquote></details>

Looks perfect.

---

*[multibody/multibody_tree/multibody_forces.h, line 79 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2f2V3IMzU_7xPQ3FUu:-L2f2V3IMzU_7xPQ3FUv:b-css0a8) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.h#L79)):*
> ```objc
> 
>   /// Adds into `this` the force contribution stored in `addend`.
>   void AddInForces(const MultibodyForces<T> &addend);
> ```

BTW Something messed up your & alignment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2f2nw64Ju8HyYmV361:b-7qm1x4)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
+@sherm1 (assigning for platform review)

---

Review status: 9 of 12 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2f2tktrmldG7i2M8vH:b-pg7j4g)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review comments:
- Looks very nice!
- The remaining uses of "forcing" seem awkward -- better to convert to "forces" everywhere IMO.
- A lot of the code here will shrink and look nicer when CalcInverseDynamics takes a MultibodyForces struct.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending BTWs below, ptal.

---

Reviewed 15 of 16 files at r4.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[examples/multibody/pendulum/pendulum_plant.cc, line 257 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2ftCUBN6m8-igjaApK:-L2ftCUBN6m8-igjaApL:b-uyptww) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/examples/multibody/pendulum/pendulum_plant.cc#L257)):*
> ```cpp
>   // Forces:
>   MultibodyForces<T> forcing(*model_);
>   // Bodies's accelerations, ordered by BodyNodeIndex.
> ```

BTW plural possessive here is `Bodies'` with no final `s`. A less-ugly alternative would be "Acceleration of the bodies, ordered by ...".

---

*[multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h, line 306 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gVAlibdSwhV56jhhk:-L2gVAlibdSwhV56jhhl:b-ws9ek8) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/benchmarks/kuka_iiwa_robot/drake_kuka_iiwa_robot.h#L306)):*
> ```objc
>     model_->CalcAccelerationKinematicsCache(*context_, pc, vc, qDDt, &ac);
> 
>     // Applied forcing:
> ```

BTW forcing -> forces (and below)

---

*[multibody/multibody_tree/force_element.h, line 63 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gO06TFVEQE-1PBq-f:-L2gO06TFVEQE-1PBq-g:bvn1dus) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/force_element.h#L63)):*
> ```objc
>   ///   This method will abort if the `forcing` pointer is null or if the
>   ///   forcing object is not compatible with `this` %MultibodyTree, see
>   ///   MultibodyForcing::CheckInvariants().
> ```

BTW "forcing" -> "forces" above in several places (probably better below too).

---

*[multibody/multibody_tree/force_element.h, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gOB4DuNqeE45McUWI:-L2gOB4DuNqeE45McUWJ:bhullhp) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/force_element.h#L74)):*
> ```objc
>                                    MultibodyForces<T>* forcing) {
>     DRAKE_DEMAND(forcing != nullptr);
>     DRAKE_DEMAND(forcing->CheckInvariants(this->get_parent_tree()));
> ```

BTW do you want `DRAKE_ASSERT` here? CheckInvariants() methods are often expensive and this gets called several times per time step.

---

*[multibody/multibody_tree/force_element.h, line 87 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gPkKmQKz2xe0vJHAK:-L2gPkKmQKz2xe0vJHAL:brp9o3h) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/force_element.h#L87)):*
> ```objc
>   /// Non-conservative force elements need to implement the rate of change,
>   /// power, at which they either generate or dissipate energy in the system in
>   /// CalcNonConservativePower().
> ```

BTW consider mentioning that a single force element may have both conservative and non-conservative contributions, such as a spring-damper element.

---

*[multibody/multibody_tree/force_element.h, line 107 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gPyDGFcltENrKBpXm:-L2gPyDGFcltENrKBpXn:b-u1xwzt) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/force_element.h#L107)):*
> ```objc
>   /// call to CalcPositionKinematicsCache().
>   /// @pre The velocity kinematics `vc` must have been previously updated with a
>   /// call to CalcVelocityKinematicsCache().
> ```

BTW slightly odd to mention this here since there is no `vc` argument.

---

*[multibody/multibody_tree/force_element.h, line 163 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gQRyq8Qq6o5ZK4wX1:-L2gQRyq8Qq6o5ZK4wX2:b-rnn8ra) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/force_element.h#L163)):*
> ```objc
>   /// parameters of this method.
>   /// It assumes `forcing` to be a valid pointer to a MultibodyForcing object
>   /// compatible with the MultibodyTree model owning `this` force element.
> ```

BTW forcing -> forces above and below.

---

*[multibody/multibody_tree/multibody_forces.cc, line 16 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gj61t2TElAGQtaHdQ:-L2gj61t2TElAGQtaHdR:b-dzuerv) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.cc#L16)):*
> ```cpp
>   DRAKE_DEMAND(model.topology_is_valid());
>   F_B_W_.resize(model.get_num_bodies(), SpatialForce<T>::Zero());
>   tau_.resize(model.get_num_velocities());
> ```

BTW this is initializing F_B_W_ to zero but doesn't look like tau_ is zeroed. Should zero them both.

---

*[multibody/multibody_tree/multibody_forces.cc, line 36 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2glA5G0Vfe93UAq-h3:-L2glA5G0Vfe93UAq-h4:b-h0q2xt) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.cc#L36)):*
> ```cpp
> template <typename T>
> void MultibodyForces<T>::AddInForces(
>     const MultibodyForces<T> &addend) {
> ```

BTW `&` placement

---

*[multibody/multibody_tree/multibody_forces.cc, line 41 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2glGCdeAve6GCq-WCt:-L2gjwukedRj9dzfd53g:b-qg5nju) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.cc#L41)):*
> ```cpp
>       F_B_W_.begin(), F_B_W_.end(),
>       addend.body_forces().begin(),
>       F_B_W_.begin(), std::plus<SpatialForce<T>>());
> ```

FYI slick, but consider this alternative:
```c++
auto next = addend.cbegin();
for (auto& force : F_B_W_)
  force += *next++;
```
That's more readable IMO (some might disagree though). More importantly it doesn't hide the fact that the result goes back into one of the inputs. That's not clear in std::transform and means it has to use operator+ rather than operator+=. Probably doesn't matter much if you prefer the std::transform.

---

*[multibody/multibody_tree/multibody_forces.h, line 16 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2ftl77EcvCyT6L1X6u:-L2ftl77EcvCyT6L1X6v:b-oivw6y) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.h#L16)):*
> ```objc
> 
> /// A class to hold a set of forces applied to a MultibodyTree system.
> /// Forces can include generalized forces as well as spatial body forces.
> ```

BTW consider "body spatial forces" rather than "spatial body forces"

---

*[multibody/multibody_tree/multibody_forces.h, line 48 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gdGLy4jP-wSpfuWie:-L2gdGLz4V-d0V1QzHuw:bhqx9wg) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.h#L48)):*
> ```objc
> 
>   /// Returns the number of generalized velocities for the model to which these
>   /// forces apply.
> ```

BTW mention that the number of generalized forces is always the same as the number of generalized velocities.

---

*[multibody/multibody_tree/multibody_forces.h, line 60 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gdYKcAKrojGiOpCza:-L2gdYKcAKrojGiOpCzb:b-ppwxol) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.h#L60)):*
> ```objc
>   }
> 
>   /// Mutable version of generalized_forces().
> ```

BTW consider a reminder here that forces should be "+="ed (added) into this array rather than just assigned (assuming that's correct). Maybe suggest users consider AddInForces() instead?

---

*[multibody/multibody_tree/multibody_forces.h, line 73 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2geDynbJ-8R8C1ddpQ:-L2geDynbJ-8R8C1ddpR:birqzho) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.h#L73)):*
> ```objc
>   }
> 
>   /// Mutable version of body_forces().
> ```

BTW consider reminder about +=

---

*[multibody/multibody_tree/multibody_forces.h, line 85 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gorl1RRQzxWEatCdi:-L2gorl1RRQzxWEatCdj:b-rx70c8) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_forces.h#L85)):*
> ```objc
>   /// @returns true if `this` forces object has the proper sizes for the given
>   /// `model`.
>   bool CheckInvariants(const MultibodyTree<T>& model) const;
> ```

BTW consider a more straightforward name like `CheckHasRightSizeForModel()`.

---

*[multibody/multibody_tree/multibody_tree.cc, line 397 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2go-1622OgIw26a5iv:-L2go-1622OgIw26a5iw:b-5rsowo) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_tree.cc#L397)):*
> ```cpp
>     const PositionKinematicsCache<T>& pc,
>     const VelocityKinematicsCache<T>& vc,
>     MultibodyForces<T>* forcing) const {
> ```

BTW forces

---

*[multibody/multibody_tree/multibody_tree.h, line 908 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2goAOYomrktXCPDxKF:-L2goAOYomrktXCPDxKG:btu31nv) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/multibody_tree.h#L908)):*
> ```objc
>   ///   objects in the model. This method will abort if the `forcing` pointer is
>   ///   null or if the forcing object is not compatible with `this`
>   ///   %MultibodyTree, see MultibodyForcing::CheckInvariants().
> ```

BTW Forcing -> Forces (above and below)

---

*[multibody/multibody_tree/uniform_gravity_field_element.h, line 38 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gvg9nLM6ReFpTduD3:-L2gvg9nLM6ReFpTduD4:b-xrhkhc) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/uniform_gravity_field_element.h#L38)):*
> ```objc
>   const Vector3<double>& gravity_vector() const { return g_W_; }
> 
>   T CalcPotentialEnergy(
> ```

BTW potential energy has an arbitrary zero definition for a field. Should mention how zero is defined here -- appears to be when a body COM is at the same height as the world origin.

---

*[multibody/multibody_tree/uniform_gravity_field_element.h, line 57 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2goM-oa1p-ZYRYMN-a:-L2goM-oa1p-ZYRYMN-b:b-5rsowo) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/uniform_gravity_field_element.h#L57)):*
> ```objc
>       const PositionKinematicsCache<T>& pc,
>       const VelocityKinematicsCache<T>& vc,
>       MultibodyForces<T>* forcing) const final;
> ```

BTW forces

---

*[multibody/multibody_tree/test/multibody_forces_test.cc, line 26 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2goXD7i6atvTfkfYit:-L2goXD7i6atvTfkfYiu:bv48cn3) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/test/multibody_forces_test.cc#L26)):*
> ```cpp
> constexpr double kEpsilon = std::numeric_limits<double>::epsilon();
> 
> class MultibodyForcingTests : public ::testing::Test {
> ```

BTW Forcing -> Forces

---

*[multibody/multibody_tree/test/multibody_forces_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gpT6cgZTH_bwknPYT:-L2gpT6cgZTH_bwknPYU:b-saklfb) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/test/multibody_forces_test.cc#L54)):*
> ```cpp
>   EXPECT_EQ(forcing.num_velocities(), model_.get_num_velocities());
> 
>   EXPECT_TRUE(forcing.generalized_forces() == Vector2d::Zero());
> ```

BTW I suspect this is just passing by luck since the constructor doesn't zero tau_. Would be a better test if we could be sure the memory was dirty prior to the alleged zeroing. One way to do that would be with placement new, something like this:
```c++
std::array<char, sizeof(MultibodyForces<double>)> mem;
mem.fill(1);  // Put a 1 in every byte.
// Default construct into mem.
auto forces = new(&mem) MultibodyForces<double>; 
// Now check forces->generalized_forces() and ->body_forces() for zero.
```
Currently I *think* this will fail with the current constructor, though I didn't actually try it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2gVzaKskRqXhbeUYz-:b-fbw8gi)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[multibody/multibody_tree/force_element.h, line 74 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gOB4DuNqeE45McUWI:-L2uzz_SK5m4SuntMBLo:b1k2ico) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/force_element.h#L74)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW do you want `DRAKE_ASSERT` here? CheckInvariants() methods are often expensive and this gets called several times per time step.
</blockquote></details>

BTW. In general that is true yes, but in particular for `MultibodyForces` the check is quite cheap.

---

*[multibody/multibody_tree/test/multibody_forces_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gpT6cgZTH_bwknPYT:-L2uyR_yNkyqXJy18Gvw:b-t0sydm) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/test/multibody_forces_test.cc#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I suspect this is just passing by luck since the constructor doesn't zero tau_. Would be a better test if we could be sure the memory was dirty prior to the alleged zeroing. One way to do that would be with placement new, something like this:
```c++
std::array<char, sizeof(MultibodyForces<double>)> mem;
mem.fill(1);  // Put a 1 in every byte.
// Default construct into mem.
auto forces = new(&mem) MultibodyForces<double>; 
// Now check forces->generalized_forces() and ->body_forces() for zero.
```
Currently I *think* this will fail with the current constructor, though I didn't actually try it.
</blockquote></details>

BTW. Good call Sherm. Fixed the construction.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Great - thanks, @amcastro-tri. Two minor BTWs added but still <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>.

---

Reviewed 3 of 16 files at r4, 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/multibody_forces.h, line 86 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2v9QNzSi9B2C3WfwT4:-L2v9QNzSi9B2C3WfwT5:bpfwl2f) ([raw file](https://github.com/robotlocomotion/drake/blob/ca2fed38ddda3d8aa5c1c9fa9977412186a18c44/multibody/multibody_tree/multibody_forces.h#L86)):*
> ```objc
>   /// @returns true if `this` forces object has the proper sizes for the given
>   /// `model`.
>   bool CheckHasRightSizeForModel(const MultibodyTree<T> &model) const;
> ```

BTW `&` moved over

---

*[multibody/multibody_tree/test/multibody_forces_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gpT6cgZTH_bwknPYT:-L2v9zIY9EXCKnCoAqR0:b-8ra20n) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/test/multibody_forces_test.cc#L54)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW. Good call Sherm. Fixed the construction.
</blockquote></details>

BTW Didn't like my hard-core test? The one here won't fail if some future programmer messes up the zeroing constructor.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-:-L2vAZL4u9BrGNABqzGd:b-x3zyqg)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 10 of 12 files reviewed at latest revision, 1 unresolved discussion.

---

*[multibody/multibody_tree/test/multibody_forces_test.cc, line 54 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7731#-L2gpT6cgZTH_bwknPYT:-L2za5M9O1AFAkWto6h4:b-gref5z) ([raw file](https://github.com/robotlocomotion/drake/blob/8944a151b02aa993c89b0520e1b1eee28cdeb019/multibody/multibody_tree/test/multibody_forces_test.cc#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW Didn't like my hard-core test? The one here won't fail if some future programmer messes up the zeroing constructor.
</blockquote></details>

BTW. I thought it was cute yes, I just didn't wanna spend time figuring out what happened to the address sanitizer. The problems was that I need to explicitly call the object's destructor or otherwise the sanitizer rightfully so complained about not releasing the internal data members of it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7731)*
<!-- Sent from Reviewable.io -->
